### ISSUE TITLE
Problem with some class annotations that cause PhpStorm mark file in error

### ISSUE THREAD

**Issue by wpottier** [Haehnchen/idea-php-annotation-plugin#30]
Using the library BazingaHateoas we have some annotation to define relations between objects. I cannot figure out why these annotations cause PHPStorm (with the PHP-Annotation plugin) warn me about an expected DOC_COMMENT_END.

The docblock: 
```php
/**
 * Class User
 *
 * @Hateoas\Relation(name="self", href=@Hateoas\Route(name="test1", parameters = { "userId"="expr(object.id)" } ))
 * @Hateoas\Relation(name="links", href=@Hateoas\Route(name="test2", parameters = { "userId"="expr(object.id)" } ), exclusion = @Hateoas\Exclusion(groups={"Own"}))
 *
 */
class Test
{
...
}
```

The error in PHPStorm display a red underline just after the end of the docblock.

Thanks for your help.

**Comment by jimigrunge** [comment]
Having same issue. Looks like it may have to do with nested Annotations and arrays. If I remove the first nested ORM\AttributeOverride for "email" the warning goes away.

    /**
     * @ORM\Table(name="auth.users")
     * @ORM\Entity(repositoryClass="HarmonyBundle\Repository\Auth\UserRepository")
     *
     * @UniqueEntity(fields="usernameCanonical", errorPath="username", message="fos_user.username.already_used")
     * @ORM\AttributeOverrides({
     *      @ORM\AttributeOverride(name="email", column=@ORM\Column(type="string", name="email", length=255, unique=false, nullable=true)),
     *      @ORM\AttributeOverride(name="emailCanonical", column=@ORM\Column(type="string", name="email_canonical", length=255, unique=false, nullable=true))
     * })
     */

**Comment by martinjeannot** [comment]
Same here, it definitely has something to do with nested annotations. Interestingly enough and as noticed by jimigrunge, **one** nested annotation does not trigger the warning (multiple will).

```php
/**
 * @Entity(repositoryClass="AffranchissementLiberte\Repository\StampRepository")
 * @Table(name="AflStamp", uniqueConstraints={@UniqueConstraint(name="code_idx", columns={"code"})})
 *
 * @XmlRoot("afl-stamp")
 * @Relation("self", href="expr('/api/v1/afl-stamps/' ~ object.getId())")
 * @Relation("operation", href="expr('/api/v1/afl-operations/' ~ object.getOperation().getId())")
 * @Relation("appointment", href="expr('/mobile-rest/appointment/' ~ object.getAppointment().getId())", exclusion=@Exclusion(excludeIf="expr(null === object.getAppointment())"))
 * @Relation("submission", href="expr('/api/v1/afl-stamps/' ~ object.getId() ~ '/submit')", attributes={"method"="POST"}, exclusion=@Exclusion(excludeIf="expr(constant('STAMP_STATUS_CREATED') != object.getStatus())"))
 * @Relation("validation", href="expr('/api/v1/afl-stamps/' ~ object.getId() ~ '/validate')", attributes={"method"="POST"}, exclusion=@Exclusion(excludeIf="expr(constant('STAMP_STATUS_SUBMITTED') != object.getStatus())"))
 * @Relation("invalidation", href="expr('/api/v1/afl-stamps/' ~ object.getId() ~ '/invalidate')", attributes={"method"="POST"}, exclusion=@Exclusion(excludeIf="expr(constant('STAMP_STATUS_SUBMITTED') != object.getStatus())"))
 */
class Stamp
{
...
}
```

**Comment by SergSlon** [comment]
+ 1

**Comment by thundo** [comment]
Same for me with a nested annotation and PhpStorm 9.0.2 with  PHP annotations 2.6.1.

The faulty snippet is

    /**
     * ...
     * @ORM\AttributeOverrides({
     *      @ORM\AttributeOverride(name="emailCanonical", column=@ORM\Column(type="string", name="email_canonical", length=255, unique=false))
     * })
     */

    class User
    {
    }

**Comment by CawaKharkov** [comment]
Same
/**
 * @ORM\Table(name="users")
 * @ORM\Entity(repositoryClass="AppBundle\Entity\UserRepository")
 * @UniqueEntity("email", groups={"register", "edit"})
 * @Loggable()
 * @ORM\AttributeOverrides({
 *      @ORM\AttributeOverride(name="username", column=@ORM\Column()),
 *      @ORM\AttributeOverride(name="usernameCanonical", column=@ORM\Column())
 * })
 */

**Comment by johndodev** [comment]
+1

![expected](https://cloud.githubusercontent.com/assets/2741956/9731275/364f498c-561c-11e5-8a68-b32a36c24103.png)

**Comment by devuo** [comment]
+1

**Comment by devuo** [comment]
Actually, never mind, if you wrap the annotation with braces, for example:

```php
/**
 * @Something("else", when={@This("is valid")})
 */
```

The plugin won't report the DOC_COMMENT_END error and everything will still work as expected.

**Comment by sh41** [comment]
@CawaKharkov  A work around for this is to change where the `{` are in the annotation. As per your example this gives DOC_COMMENT_END error in PHPStorm
```php
/**
 * Class User
 * @package AppBundle\Entity
 * @ORM\Entity()
 * @UniqueEntity(fields="usernameCanonical", errorPath="email", message="fos_user.email.already_used")
 * @ORM\AttributeOverrides({
 * @ORM\AttributeOverride(name="email", column=@ORM\Column(type="string", name="email", length=255, unique=false, nullable=false)),
 * @ORM\AttributeOverride(name="emailCanonical", column=@ORM\Column(type="string", name="email_canonical", length=255, unique=false, nullable=false))
 * })
 *
 * @ExclusionPolicy("none")
 *
 */
class User extends BaseUser
```

This does not:
```php
/**
 * Class User
 * @package AppBundle\Entity
 * @ORM\Entity()
 * @UniqueEntity(fields="usernameCanonical", errorPath="email", message="fos_user.email.already_used")
 * @ORM\AttributeOverrides(
 * {@ORM\AttributeOverride(name="email", column=@ORM\Column(type="string", name="email", length=255, unique=false, nullable=false)),
 * @ORM\AttributeOverride(name="emailCanonical", column=@ORM\Column(type="string", name="email_canonical", length=255, unique=false, nullable=false))
 * })
 *
 * @ExclusionPolicy("none")
 *
 */
class User extends BaseUser
```

The only change is moving `{` from the end of the `@ORM\AttributeOverrides(` line to the start of the `@ORM\AttributeOverride(name="email"...`  line.

Hope that helps.

**Comment by Haehnchen** [comment]
that's not an issue here, document structure is delieverd by phpstorm. will open an external issue on jetbrains side

**Comment by sh41** [comment]
Thanks Daniel, your work on the plugin is much appreciated!

**Comment by Alumbrados** [comment]
+1

**Comment by preemiere** [comment]
+1

**Comment by githoober** [comment]
+1

**Comment by zazoomauro** [comment]
+1

**Comment by enekochan** [comment]
I was having this issue too. I had a QueryParam annotation and inside of it a Date constraint annotation. Only when 2 or more of those were used in the same PHPDoc section gave the error.

So as @sh41 suggested I changed this:

``` php
use Symfony\Component\Validator\Constraints as Assert;

    /**
     * Collection get action
     *
     * @ParamConverter("business", class="AppBundle:Business")
     * @Security("is_granted('VIEW', business)")
     * @Rest\QueryParam(name="date", strict=true, nullable=true, allowBlank=true, description="Specific date for appointments list", requirements=@Assert\Date())
     * @Rest\QueryParam(name="dateFrom", strict=true, nullable=true, allowBlank=true, description="Specific date for appointments list", requirements=@Assert\Date())
     * @Rest\QueryParam(name="dateTo", strict=true, nullable=true, allowBlank=true, description="Specific date for appointments list", requirements=@Assert\Date())
     * @Rest\QueryParam(name="pending", strict=true, nullable=true, allowBlank=true, description="Show only pending appointments", requirements="(false|true)")
     * @Rest\View()
     * @ApiDoc(
     *  authentication=true,
     *  section="Business Appointments",
     *  description="Get all appointments for a business",
     *  statusCodes={
     *      200="Returned when successful"
     *  }
     * )
     *
     * @param \AppBundle\Entity\Business $business Business
     * @param \FOS\RestBundle\Request\ParamFetcher $paramFetcher ParamFetcher
     *
     * @return \FOS\RestBundle\View\View
     */
```

To this and the error was gone:

``` php
use Symfony\Component\Validator\Constraints as Assert;

    /**
     * Collection get action
     *
     * @ParamConverter("business", class="AppBundle:Business")
     * @Security("is_granted('VIEW', business)")
     * @Rest\QueryParam(
     *     name="date",
     *     strict=true,
     *     nullable=true,
     *     allowBlank=true,
     *     description="Specific date for appointments list",
     *     requirements=@Assert\Date()
     * )
     * @Rest\QueryParam(
     *     name="dateFrom",
     *     strict=true,
     *     nullable=true,
     *     allowBlank=true,
     *     description="Specific date from for appointments list",
     *     requirements=@Assert\Date()
     * )
     * @Rest\QueryParam(
     *     name="dateTo",
     *     strict=true,
     *     nullable=true,
     *     allowBlank=true,
     *     description="Specific date to for appointments list",
     *     requirements=@Assert\Date()
     * )
     * @Rest\QueryParam(name="pending",
     *     strict=true, nullable=true,
     *     allowBlank=true,
     *     description="Show only pending appointments",
     *     requirements="(false|true)"
     * )
     * @Rest\View()
     * @ApiDoc(
     *  authentication=true,
     *  section="Business Appointments",
     *  description="Get all appointments for a business",
     *  requirements={
     *      {
     *          "name"="business",
     *          "dataType"="integer",
     *          "requirement"="\d+",
     *          "description"="Business id"
     *      }
     *  },
     *  statusCodes={
     *      200="Returned when successful"
     *  }
     * )
     *
     * @param \AppBundle\Entity\Business $business Business
     * @param \FOS\RestBundle\Request\ParamFetcher $paramFetcher ParamFetcher
     *
     * @return \FOS\RestBundle\View\View
     */
    public function getAppointmentsAction(Business $business, ParamFetcher $paramFetcher) {
        ....
    }
```

So only adding some carrier returns fixed it. Hope it helps

**Comment by Haehnchen** [comment]
fixed by #55 and PhpStorm `2016.1.2`
